### PR TITLE
Add warning about Datasharing limitation with Datareader's history size [11926]

### PR DESCRIPTION
**Pull request by richiware**
There is a limitation about Datareader's HistoryQos and Datasharing. In case it is detected, a warning is shown to the user.

### CODE REVIEW

File: src/cpp/rtps/DataSharing/DataSharingListener.cpp [hunk 1]
```diff
@@ -217,11 +218,21 @@ bool DataSharingListener::add_datasharing_writer(
 
     std::shared_ptr<ReaderPool> pool =
             std::static_pointer_cast<ReaderPool>(DataSharingPayloadPool::get_reader_pool(is_volatile));
-    pool->init_shared_memory(writer_guid, datasharing_pools_directory_);
-    writer_pools_.emplace_back(pool, pool->last_liveliness_sequence());
-    writer_pools_changed_.store(true);
+    if (pool->init_shared_memory(writer_guid, datasharing_pools_directory_))
+    {
+        if (0 == reader_history_max_samples ||
+                reader_history_max_samples > static_cast<int32_t>(pool->history_size() - 1))
```
Comment (MiguelCompany) [inline src/cpp/rtps/DataSharing/DataSharingListener.cpp:224]: Negative values also mean infinite.
`A > (B - 1)` ==> `A >= B`

```suggestion
        if (0 <= reader_history_max_samples ||
                reader_history_max_samples >= static_cast<int32_t>(pool->history_size()))
```

File: src/cpp/rtps/DataSharing/DataSharingListener.cpp [hunk 2]
```diff
@@ -217,11 +218,21 @@ bool DataSharingListener::add_datasharing_writer(
...
+        {
+            logWarning(RTPS_READER,
+                    "Reader " << reader_->getGuid() << " was configured to have a large history (" << reader_history_max_samples << " max samples), but the history size used with writer " << writer_guid << " will be " << pool->history_size() <<
```
Comment (MiguelCompany) [inline src/cpp/rtps/DataSharing/DataSharingListener.cpp:227]: I think the linter will complain about this long line

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: src/cpp/rtps/DataSharing/DataSharingListener.cpp (hunk 1 above)
Comment (richiware) [reply src/cpp/rtps/DataSharing/DataSharingListener.cpp:224]:
```suggestion
        if (0 >= reader_history_max_samples ||
                reader_history_max_samples >= static_cast<int32_t>(pool->history_size()))
```

Comment (richiware) [reply src/cpp/rtps/DataSharing/DataSharingListener.cpp:224]: Should be in this way to catch infinitive values?

Comment (MiguelCompany) [reply src/cpp/rtps/DataSharing/DataSharingListener.cpp:224]: The first condition is for the case where `reader_history_max_samples` is infinite. I'm assuming that a negative value also means infinite.

On the second condition, I think that `pool->history_size()` will always return a strictly positive value, as it is the way the pool is designed to be.

Comment (MiguelCompany) [reply src/cpp/rtps/DataSharing/DataSharingListener.cpp:224]: Now I see what you mean! It is `0 >= reader_history_max_samples`, not `0 <= reader_history_max_samples` !

**Review by MiguelCompany — APPROVED** [review]
LGTM

**Comment by richiware** [comment]
Tests were passed in #2047.

**Pull request merged**